### ISSUE TITLE
Feature request, flags to prevent up/downgrading packages on incremental installations

### ISSUE THREAD

**Issue by stuartarchibald** [conda/conda#7688]
We (Numba devs) keep getting hit by the following pattern in the giant number of configurations Numba is tested on. Essentially, incrementally building up conda envs can lead to unexpected downgrades etc which are not desirable in a testing environment (we test with `conda build` too, but this checks the path of users installing stuff manually).

As a motivating example, when testing Python 3.7, we set up envs based on the test channel for Python 3.7, then installed e.g. SciPy for testing, but as there was no SciPy+Python 3.7 build at the time, Python itself got downgraded to 3.6 so SciPy would install. As a result we were then testing on something we didn't expect to be and it e.g. erroneously passed.

One suggestion that might help with this is to add a flag to be able to do e.g. `conda install foo --downgrade-is-error` to install `foo` into an existing env, and if as a result of installing `foo` things that already exist in the env need downgrading, that's an error. The inverse flag `--upgrade-is-error` does the same but errors if installing `foo` would cause upgrades. These flags also should be able to be used simultaneously.

Thanks for considering this.

**Comment by mingwandroid** [comment]
Those some like a useful addition

**Comment by drdavella** [comment]
I agree that this would be very useful. The absence of these options can lead to a lot of headaches in test automation scenarios, as @stuartarchibald said above.

I wonder how difficult this would be to implement, and whether you would accept outside contributions for such a change? I might be able to spare a few hours for this if that's all it would require.

**Comment by kalefranz** [comment]
There are a couple of different ways it could be implemented.  But before we jump into implementation, I'd like to spec out exactly what we want in the design. And also affirm that it's really needed.

In the example case that @stuartarchibald describes, I'd argue that the correct setup would have been to make sure that `python=3.7` is a command-line input in the CI case where you're using the `--yes` flag or `always_yes` config parameter.  It seems that would be an alternate solution to the problem, that also requires no additional implementation or further expansion of all of the various CLI flags/config parameters, no?

**Comment by drdavella** [comment]
@kalefranz thanks, I actually think that's a good enough workaround for me, although I'd still argue that having an explicit option might be a little cleaner/more intuitive in the long run.

**Comment by kalefranz** [comment]
I'm starting to hear low-level rumbling of "configuration paralysis" with all the CLI flags and config parameters.  So just being careful that we actually need to implement something new :)

**Comment by stuartarchibald** [comment]
Thanks @kalefranz, that may be ok as a work around, I'll give it a try, it does mean carrying around this info to all the conda install commands but so be it. As already noted, this is not necessarily intuitive although it probably produces the desired result, perhaps this situation could be added into a FAQ or similar and we'll see how things go in our (Numba) test system? Thanks again for your suggestion.

**Comment by stuartarchibald** [comment]
@kalefranz hmmm, I'm beginning to think that this is might not always work a work around. We (Numba team) have just been caught out again in CI testing by an issue where we do approximately the following:
```
conda create -n foo python=2.7 numpy=1.15 pip # we get python=2.7.15
conda install scipy # due to some dependency problem, this causes the python 2.7.15 installed to downgrade to 2.7.3
```
Real logs are here: https://dev.azure.com/numba/numba/_build/results?buildId=211&view=logs it's the OSX builder that has the issue.

Short of parsing the output of conda commands (json or not) to push initially installed version numbers into future install commands as contraints... can the notion of "install the latest builds of X, Y, Z" followed by "additionally install A and B but do not downgrade X, Y, Z in any way at all" be achieved?

**Comment by tjd2002** [comment]
> Short of parsing the output of conda commands (json or not) to push initially installed version numbers into future install commands as contraints... can the notion of "install the latest builds of X, Y, Z" followed by "additionally install A and B but do not downgrade X, Y, Z in any way at all" be achieved?

I'm surprised no one has suggested pinning yet, as it seems to be very well suited to the CI problem at hand. The following solution does involve parsing the output of `conda list`, but it's fairly straightforward, and is a one-time operation (it does not require that you remember to pass the full list of constraints to all subsequent incremental installs).

```
conda create -n foo python=2.7 numpy=1.15 pip # installs python=2.7.15
conda activate foo  # [1]
conda list --full --export '(python|numpy)' | grep -v "^#" | xargs -t -n 1 conda config --env --add pinned_packages # [2]
conda install scipy # will fail if it requires python < 2.7.15, 
conda install python=2.7.3  # fails
```

[1] You have to activate the env so that you can use the `--env` flag in `conda config`

[2] The grep is needed to remove the comment lines that conda list adds to the output when using `--export`. (The output of `conda list -c` won't work here as canonical path names are not valid MatchSpec's for pinning. )

Note that this solution will also fail if subsequent install steps result in package *upgrades*. If you don't want this you could do some munging to replace the first `=` in the MatchSpec with `>=`.

**Comment by kalefranz** [comment]
I haven’t yet validated this myself (I only have my phone right now), but I suspect the below is true

```
conda create -n foo python=2.7.15 numpy=1.15 pip # installs python=2.7.15
conda activate foo  # [1]
conda install scipy # will fail if it requires python < 2.7.15
```

The only difference is that the python spec is implicitly pinned at `2.7.15` instead of just `2.7` as a user-requested spec.

Conda keeps a record of all specs passed in through the command line as “user-requested” specs. The full history of user requests is considered for each solve. A spec passes via the command line is a “hard” requirement for the current solve, and in subsequent solves becomes a softer form of pinning. Softer, because a subsequent spec for the same package name can override a previously-set spec. For example, extending from the above scenario

```
conda install scipy # will fail if it requires python < 2.7.15
conda install scipy python=2.7 # should succeed by letting python float up to 2.7.15 or later
```

**Comment by kalefranz** [comment]
Just reading back over this whole issue, I want to make it clear that I support the feature request and think it’s an excellent idea in general (I’ve actually wanted to use these flags myself over the last 6-9 months). The purpose of my comments in this issue was to share and clarify what already exists, so that we don’t build competing ways to accomplish the same thing.

**Comment by tjd2002** [comment]
@kalefranz This is what I thought 'soft-pinning' meant, but I don't see it working as expected, as in the following example:

```
conda create -n py37 python=3.7
conda install -n py37 pysam # a package that is only available for python 2.7
```
The solver happily proposes downgrading python to 2.7, even though I specified 3.7 on the command line. 

I'm on conda 4.6.14. Am I misunderstanding this functionality, or should I file a separate issue?

**Comment by stuartarchibald** [comment]
I presume that this behaviour is still present, but seem to recall that major point versions of Python will now not be changed unless permitted by a flag?
